velox/exec/tests/RowContainerTest.cpp

```diff
@@ -47,29 +74,102 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
 
     auto result = BaseVector::create(expected->type(), size, pool_.get());
     container.extractColumn(oddRows.data(), size, column, result);
-    EXPECT_EQ(size, result->size());
-    for (vector_size_t row = 0; row < size; ++row) {
-      if (row % 2 == 0) {
-        EXPECT_TRUE(result->isNullAt(row)) << "at " << row;
+    return result;
+  }
+
+  VectorPtr testExtractColumnRowNumbersForOddRows(
```

Comment:
I find test logic confusing and difficult to reason about. I'm seeing that testExtractColumnForOddRows and testExtractColumnForAllRows are always called together with the same arguments. I suggest to introduce testExtractColumn method that combines the two and update the call sites.

testExtractColumn then can call a few helper methods to test different ways of extracting a column.

```
void testExtractColumn() {
   /// Extract all rows without result offset.
   /// Extract all rows with result offset.
   /// Extract odd rows without result offset.
   /// Extract odd rows with result offset.
   /// Extract individual rows (specified by row number) with result offset.
}
```

I'd modify existing testExtractColumnFor[All/Odd]Rows to add optional resultOffset argument. Then, add a new testExtractColumnUsingRowNumbers.

Reply:
I've refactored the tests more clean now. Please give me your comments on them.